cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/batch/BatchEmitter.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.api.batch;
+
+import co.cask.cdap.api.annotation.Beta;
+import co.cask.cdap.etl.api.Emitter;
+
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Batch emitter which is used to emit records one by one to next stages
+ * @param <T> type of transform detail
+ */
+@Beta
+public abstract class BatchEmitter<T> implements Emitter<Object> {
+
+  public abstract void addTransformDetail(String stageName, T transformDetail);
+
+  @Nullable
```

Comment:
code is cleaner if you don't have to worry about nulls. Return an empty map if there are no next stages.